### PR TITLE
Add an Eventbus to send messages

### PR DESCRIPTION
**Pull request by jverbraeken**
- Relevant Issues: -
- Related Pull Requests: -

* * *

## What
This PR makes it possible to send messages between modules to communicate

## Why
You want to send message from a module to another module, for example you'll want to notify all modules when they can start using the service locator, you want to let the UI know when a new Bluetooth device is detected and you'll want to communicate with other modules recently received public keys.

## How
This PR is implemented by using Greenrobot's Eventbus, arguably the best open-source Eventbus library optimized for Android

## Alternative implementation
I have considered to implement an Eventbus myself, but Greenrobot's eventbus is proven to work and has some advanced features like support for multithreading.

## Notes
-

### CODE REVIEW

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java [hunk 1]
```diff
@@ -29,4 +35,14 @@ private EncryptorAdapter(final IServiceLocatorCreator serviceLocatorCreator) {
     public static ModuleWrapper<EncryptorAdapter> newInstance(final IServiceLocatorCreator serviceLocatorCreator) {
         return new ModuleWrapper<>(new EncryptorAdapter(serviceLocatorCreator));
     }
+
+    /**
+     * {@inheritDoc}
+     * @param event Indicates that the {@link SLReadyEvent} happened
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java]: Not needed when `{@inheritDoc}` is used (goes for other classes as well)

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorCreator.java [hunk 2]
```diff
@@ -38,7 +41,9 @@ private ServiceLocatorCreator() {
      * @return An interface containing the methods that may be used on the newly constructed {@link IServiceLocator}
      */
     static IServiceLocator newInstance() {
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorCreator.java]: Wouldn't `createInstance` be a more descriptive name?

File: app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorCreator.java [hunk 3]
```diff
@@ -38,7 +41,9 @@ private ServiceLocatorCreator() {
...
-        return new ServiceLocatorCreator().serviceLocator;
+        final IServiceLocator tmp = new ServiceLocatorCreator().serviceLocator;
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/service_locator/ServiceLocatorCreator.java]: Please use a more descriptive name for this variable

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
+package com.nervousfish.nervousfish;
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: This shouldn't be part of this PR

File: app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java [hunk 5]
```diff
@@ -1,7 +1,15 @@
 package com.nervousfish.nervousfish;
```
Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/ExampleUnitTest.java]: These changes shouldn't be part of this PR

File: app/src/test/java/com/nervousfish/nervousfish/BaseTest.java (hunk 4 above)
Comment (jverbraeken) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Yeah, the file is in develop already. Not sure why git sees this file a change compared to develop

Comment (ericcornelissen) [inline app/src/test/java/com/nervousfish/nervousfish/BaseTest.java]: Well, it differs from the one in develop so you should at least fix that 😉

File: app/build.gradle [hunk 6]
```diff
@@ -73,7 +73,7 @@ dependencies {
     compile 'org.greenrobot:eventbus:3.0.0'
 
     testCompile 'junit:junit:4.12'
-    testCompile 'org.mockito:mockito-core:2.7.22'
+    androidTestCompile 'org.mockito:mockito-core:2.7.22'
```
Comment (ericcornelissen) [inline app/build.gradle]: We want mockito support in both the normal tests (`testCompile`) and the android tests (`androidTestCompile`) 😉

File: app/src/androidTest/java/com/nervousfish/nervousfish/test/ExampleUnitTest.java [hunk 7]
```diff
@@ -16,8 +20,9 @@
  * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
  */
 @SuppressWarnings("PMD")
+@RunWith(AndroidJUnit4.class)
```
Comment (ericcornelissen) [inline app/src/androidTest/java/com/nervousfish/nervousfish/test/ExampleUnitTest.java]: Is this needed given [line 75](https://github.com/ericcornelissen/NervousFish/blob/99a350269917d364076cb370f4be44c730d66075/app/build.gradle#L75) of build.gradle?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyNFCHandler.java [hunk 8]
```diff
@@ -1,22 +1,28 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An handler doing nothing.
  */
 public final class DummyNFCHandler extends APairingHandler implements INFCHandler {
+    @SuppressWarnings("PMD")
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyNFCHandler.java]: Can you specify the exact thing you're suppressing?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java [hunk 9]
```diff
@@ -1,22 +1,28 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An handler doing nothing.
  */
 public final class DummyQRHandler extends APairingHandler implements IQRHandler {
+    @SuppressWarnings("PMD")
```
Comment (ericcornelissen) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyQRHandler.java]: Can you specify the exact thing you're suppressing?

File: app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java [hunk 10]
```diff
@@ -1,22 +1,28 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An handler doing nothing.
  */
 public final class DummyBluetoothHandler extends APairingHandler implements IBluetoothHandler {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/pairing/DummyBluetoothHandler.java]: What Eric said about pmd 😆

File: app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An adapter to the default Android file system
  */
 public final class AndroidFileSystemAdapter implements IFileSystem {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/filesystem/AndroidFileSystemAdapter.java]: What Eric said about pmd 😆

File: app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java [hunk 12]
```diff
@@ -1,21 +1,27 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An adapter to the Gson database library
  */
 public final class GsonDatabaseAdapter implements IDatabase {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapter.java]: Again

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java [hunk 13]
```diff
@@ -1,21 +1,27 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An adapter to the default Java class for generating keys
  */
 public final class KeyGeneratorAdapter implements IKeyGenerator {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/KeyGeneratorAdapter.java]: Again

File: app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java [hunk 14]
```diff
@@ -1,22 +1,27 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * An adapter to the default Java class for encrypting messages
  */
 public final class EncryptorAdapter implements IEncryptor {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/cryptography/EncryptorAdapter.java]: Again

File: app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java [hunk 15]
```diff
@@ -1,21 +1,26 @@
...
 import com.nervousfish.nervousfish.service_locator.IServiceLocatorCreator;
 import com.nervousfish.nervousfish.service_locator.ModuleWrapper;
 
+import org.greenrobot.eventbus.Subscribe;
+
 /**
  * Contains the main constants
  */
 public final class Constants implements IConstants {
+    @SuppressWarnings("PMD")
```
Comment (TheBonheurs) [inline app/src/main/java/com/nervousfish/nervousfish/modules/constants/Constants.java]: And again

File: app/build.gradle (hunk 6 above)
Comment (ericcornelissen) [inline app/build.gradle]: @jverbraeken You wanne have **both** `testCompile 'org.mockito:mockito-core:2.7.22'` and `androidTestCompile 'org.mockito:mockito-core:2.7.22'`...
